File: plugins/reels-factory/engine/scripts/hf_sdk.mjs
// Мост к их редактору композиций: @hyperframes/sdk.
//
// Разбор HTML и правку композиции движок держит готовыми — «headless,
// framework-neutral composition editing engine for agents and custom editors»
// (packages/sdk/package.json:4). Сессия без адаптеров и есть наш режим:
// «Headless (agents): omit both — SDK is a stateless transform + serializer»
// (packages/sdk/src/session.ts:851-852). Своего разборщика HTML у нас поэтому
// больше нет.
//
// Процесс живёт всю сборку и говорит построчным JSON: открыть композицию —
// прочитать её элементы — применить правки — записать результат. Один запуск
// node на сборку вместо запуска на блок.
//
// Протокол (по строке на сообщение):
//   {"id":1,"cmd":"open","name":"g13","path":"…/g13.html"}      -> {"ok":true}
//   {"id":2,"cmd":"elements","name":"g13"}                      -> {"ok":true,"elements":[…]}
//   {"id":3,"cmd":"dispatch","name":"g13","ops":[EditOp,…]}     -> {"ok":true}
//   {"id":4,"cmd":"save","name":"g13","path":"…/copy.html"}     -> {"ok":true,"bytes":N}
//   {"id":5,"cmd":"close","name":"g13"}                         -> {"ok":true}
// Ошибка: {"ok":false,"error":"…"}.
//
// Файлами, а не текстом: блок каталога весит под полмегабайта, и гонять его
// через канал по два раза на карточку дороже, чем прочитать с диска.
import { openComposition } from "@hyperframes/sdk";
import { parseHTML } from "linkedom";
import { readFileSync, writeFileSync } from "node:fs";
import { createInterface } from "node:readline";

const sessions = new Map();

/** Собственный текст каждого элемента — только его текстовые узлы.
 *
 * Их снимок отдаёт поле `text`, но оно смотрит «сквозь» единственного потомка
 * (`packages/sdk/src/engine/model.ts:337-345`): у `<span><i></i>СЛОВО</span>`
 * текстом числится пустая полоска подсветки, а слово теряется. В нашем
 * каталоге акцентное слово устроено ровно так, поэтому текст читаем с их же
 * разметки их же разборщиком — linkedom, зависимость их SDK. Метки
 * `data-hf-id` проставляет сам SDK (`parsers/src/hfIds.ts:5`), по ним и
 * сходятся два взгляда на один документ.
 */
function ownTexts(comp) {
  const { document } = parseHTML(comp.serialize());
  const map = new Map();
  for (const element of document.querySelectorAll("[data-hf-id]")) {
    let text = "";
    element.childNodes.forEach((node) => {
      if (node.nodeType === 3) text += node.nodeValue ?? "";
      else if (node.nodeType === 1 && node.tagName === "BR") text += "\n";
    });
    map.set(element.getAttribute("data-hf-id"), text);
  }
  return map;
}

/** Плоский список элементов в порядке документа: у каждого свои дети и место. */
function flatten(comp) {
  const texts = ownTexts(comp);
  const out = [];
  const walk = (element, parent, index) => {
    const self = out.length;
    out.push({
      hfid: element.scopedId || element.id,
      tag: element.tag,
      classes: [...element.classNames],
      attrs: { ...element.attributes },
      text: texts.get(element.id) ?? "",
      start: element.start,
      duration: element.duration,
      trackIndex: element.trackIndex,
      anims: [...element.animationIds],
      parent,
      index,
      children: [],
    });
    element.children.forEach((child, position) => {
      out[self].children.push(out.length);
      walk(child, self, position);
    });
  };
  comp.getRootElements().forEach((root, position) => walk(root, null, position));
  return out;
}

function need(name) {
  const session = sessions.get(name);
  if (!session) throw new Error(`композиция ${name} не открыта`);
  return session;
}

async function handle(message) {
  switch (message.cmd) {
    case "open": {
      const html = readFileSync(message.path, "utf8");
      // history:false — стек отмены нам не нужен, а он держит копию документа
      // на каждую правку: на блоке в полмегабайта это заметная память.
      sessions.set(message.name, await openComposition(html, { history: false }));
      return { ok: true };
    }
    case "elements":
      return { ok: true, elements: flatten(need(message.name)) };
    case "dispatch": {
      const comp = need(message.name);
      // Пачкой: их движок сворачивает правки одной транзакцией.
      comp.batch(() => {
        for (const op of message.ops) comp.dispatch(op);
      });
      return { ok: true };
    }
    case "save": {
      const html = need(message.name).serialize();
      writeFileSync(message.path, html, "utf8");
      return { ok: true, bytes: html.length };
    }
    case "close":
      need(message.name).dispose();
      sessions.delete(message.name);
      return { ok: true };
    // Кусок чужого файла целиком, разметкой: так подключается их компонент
    // субтитров — «paste its contents into your composition»
    // (docs/catalog/components/caption-highlight.mdx). Через сессию SDK это не
    // достать: снимок отдаёт разбор, а не исходную разметку, — поэтому файл
    // читается их же разборщиком напрямую.
    case "extract": {
      const { document } = parseHTML(readFileSync(message.path, "utf8"));
      const found = message.all
        ? [...document.querySelectorAll(message.selector)]
        : [document.querySelector(message.selector)].filter(Boolean);
      return {
        ok: true,
        found: found.map((el) => ({ outer: el.outerHTML, text: el.textContent })),
      };
    }
    default:
      throw new Error(`неизвестная команда ${message.cmd}`);
  }
}

const lines = createInterface({ input: process.stdin });
for await (const line of lines) {
  if (!line.trim()) continue;
  let request;
  try {
    request = JSON.parse(line);
  } catch (error) {
    process.stdout.write(JSON.stringify({ ok: false, error: String(error) }) + "\n");
    continue;
  }
  let answer;
  try {
    answer = await handle(request);
  } catch (error) {
    answer = { ok: false, error: error?.message ?? String(error) };
  }
  answer.id = request.id;
  process.stdout.write(JSON.stringify(answer) + "\n");
}

File: plugins/reels-factory/engine/src/reels_factory/hf_sdk.py
"""Их редактор композиций из питона: тонкая обёртка над `@hyperframes/sdk`.

Разбор HTML, правку элементов и запись обратно движок держит готовыми. Своего
разборщика у нас нет: элементы, их атрибуты, собственный текст и даже список
твинов, целящихся в элемент, отдаёт `comp.getElements()`
(`packages/sdk/src/types.ts:25-50`), а правки идут их же операциями
(`packages/sdk/src/types.ts:104-163`).

Процесс node поднимается один на сборку и живёт, пока открыт `sdk_session()`.
"""
from __future__ import annotations

import json
import os
import shutil
import subprocess
from contextlib import contextmanager
from dataclasses import dataclass, field
from pathlib import Path

BRIDGE = Path(__file__).resolve().parents[2] / "scripts" / "hf_sdk.mjs"

#: Чем звать node. На Windows это `node.exe` в PATH, в WSL — обычный `node`.
NODE = os.environ.get("REELS_NODE") or shutil.which("node") or "node"


@dataclass
class Node:
    """Элемент композиции в разборе их SDK."""

    hfid: str
    tag: str
    classes: list[str]
    attrs: dict[str, str]
    text: str | None
    start: float | None
    duration: float | None
    track_index: int | None
    anims: list[str]
    parent: int | None
    index: int
    children: list[int] = field(default_factory=list)

    def has_class(self, name: str) -> bool:
        return name in self.classes


def _to_node(raw: dict) -> Node:
    return Node(hfid=raw["hfid"], tag=raw["tag"], classes=raw["classes"],
                attrs=raw["attrs"], text=raw["text"], start=raw["start"],
                duration=raw["duration"], track_index=raw["trackIndex"],
                anims=raw["anims"], parent=raw["parent"], index=raw["index"],
                children=raw["children"])


class SdkError(RuntimeError):
    pass


class Sdk:
    """Открытые композиции и очередь правок к ним."""

    def __init__(self, process: subprocess.Popen):
        self._process = process
        self._id = 0

    def _call(self, **message) -> dict:
        self._id += 1
        message["id"] = self._id
        self._process.stdin.write(json.dumps(message, ensure_ascii=False) + "\n")
        self._process.stdin.flush()
        line = self._process.stdout.readline()
        if not line:
            tail = (self._process.stderr.read() or "")[-1000:]
            raise SdkError(f"мост к их SDK оборвался: {tail}")
        answer = json.loads(line)
        if not answer.get("ok"):
            raise SdkError(answer.get("error") or "мост к их SDK вернул отказ")
        return answer

    def open(self, name: str, path) -> None:
        self._call(cmd="open", name=name, path=str(path))

    def elements(self, name: str) -> list[Node]:
        return [_to_node(raw) for raw in self._call(cmd="elements", name=name)["elements"]]

    def dispatch(self, name: str, ops: list[dict]) -> None:
        if ops:
            self._call(cmd="dispatch", name=name, ops=ops)

    def save(self, name: str, path) -> None:
        self._call(cmd="save", name=name, path=str(path))

    def close(self, name: str) -> None:
        self._call(cmd="close", name=name)

    def extract(self, path, selector: str, *, all: bool = False) -> list[dict]:
        """Куски чужого файла разметкой: `{"outer": …, "text": …}` на каждый."""
        return self._call(cmd="extract", path=str(path), selector=selector,
                          all=all)["found"]


@contextmanager
def sdk_session():
    """Поднять мост на время сборки."""
    if not BRIDGE.exists():
        raise SdkError(f"нет моста {BRIDGE}")
    process = subprocess.Popen(
        [NODE, str(BRIDGE)], cwd=str(BRIDGE.parent),
        stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
        text=True, encoding="utf-8", errors="replace")
    try:
        yield Sdk(process)
    finally:
        try:
            process.stdin.close()
        except OSError:
            pass
        try:
            process.wait(timeout=10)
        except subprocess.TimeoutExpired:
            process.kill()


# ------------------------------------------------------------------ операции

def set_text(target: str, value: str) -> dict:
    return {"type": "setText", "target": target, "value": value}


def set_attribute(target: str, name: str, value: str | None) -> dict:
    return {"type": "setAttribute", "target": target, "name": name, "value": value}


def set_timing(target: str, *, start: float | None = None,
               duration: float | None = None,
               track_index: int | None = None) -> dict:
    op = {"type": "setTiming", "target": target}
    if start is not None:
        op["start"] = start
    if duration is not None:
        op["duration"] = duration
    if track_index is not None:
        op["trackIndex"] = track_index
    return op


def remove_element(target: str) -> dict:
    return {"type": "removeElement", "target": target}


def add_element(parent: str | None, index: int, html: str) -> dict:
    return {"type": "addElement", "parent": parent, "index": index, "html": html}


def remove_tween(animation_id: str) -> dict:
    return {"type": "removeGsapTween", "animationId": animation_id}
